### PR TITLE
Support for custom scopes, ESIKey AccessMask, and ESI method setup via attributes

### PR DESCRIPTION
**Pull request by wbSD**
- Added header+description attribute to most ESI Character Methods, so they can be refreshed manually and individually, and seen in the throbber menu
- Reworked the Implants/Attributes import. They are now separate CharacterQueryMonitors again, and the order in which they import doesn't cause attributes to become higher than they are. One single instance may behave weird, if a character switches from no implants to having implants, the imported implants won't affect the attributes until attributes are re-imported. Better than showing too high, I supposed.
- UpdatePeriod attributes for Character methods are now up to date with ESI cache period as the minimum period. Default period is a bit longer in most cases. One exception is Attributes which is cached for 1hr, but gave it same period as Implants (to reduce the occurrence the single issue mentioned above). Also removed the unused CacheStyle value.
- Moved the ESI path to an attribute on the ESIMethod enum values, and added required scopes
- ESIKey now has two AccessMask properties - one for Character and one for Corporation. Upon new ESI key generation, they are set according to the scopes that the key has. These new masks are used by the QueryMonitor class to check if we have access to call the API. The serializable key still has the old AccessMask for backward compatibility, the new version will treat it as having full access.
- Added a new window for configuring ESI scopes when creating/updating an ESI key. This window allows for fully customizing which scopes you want to request for the given key. There are also some predefined simple options.

### CODE REVIEW

**Comment by wbSD** [comment]
@peterhaneve I've gone the path that you mentioned, re-using the AccessMask framework from the old API keys.
I've split it up into 2 (Character and Corporation AccessMask) because the ulong values for the character and corporation method enums had some overlap.
All of the ESI method paths have been moved to attributes on the enum values, and have had Scope added where it was required. There were some unused enum values in all 3 ESIAPIxxMethods enums, I've just left them without the attribute for now.

I haven't thought of a good way for how to set up the ESI scopes per character, so suggestions are welcome. It would pretty much just be a matter of generating the CharacterAccessMask and CorporationAccessMask that matches what you want access to, build a scopes string from that, and pass it on to the SSOAuthenticationService.SpanBrowserForLogin() method or something similar.

**Comment by peterhaneve** [comment]
Nothing against your PR, but I am going to be busy for the next two weeks and may not have time to look at this in detail.

**Comment by wbSD** [comment]
I've removed all traces of the SSO scope setting from the settings page, this now only uses the new AccessMask and the scopes that the OAuth response sends back.
Let me know what you think, when you've had time to look at it.

**Comment by peterhaneve** [comment]
Wow, this change exploded quickly. It will take me a bit to go through it, but good work!

Based on the description for b524cc6 does it require people to delete/re-add all characters to make this work? I have seen use cases with 50+ 2fa accounts.

(As an aside, I had much simpler ambitions in mind and simply wanted to add a few checkboxes in the ESI key window for "Core Features" "Assets" "Corporation" "Markets&Contracts" etc. with hardcoded scope lists)

**Comment by wbSD** [comment]
Yea I may have gone a bit overboard here, but I really wanted to add the ability to control it on a per-scope basis.
The last thing I added was actually backward compatibility with the old AccesMask property, which is currently set to ulong max value. If it is that, it's just copied over as max value to the two new access masks, so people don't have to re-add everything. https://github.com/peterhaneve/evemon/pull/13/commits/6c83baaae02a33b25b1b9a36a6bde822d73c1d4d

Since I left out the checkbox part, it's still open for how to implement it. I tried to make it so that it's easy to just use the ESI method enum attributes to compile the list of scopes, no matter how it's done.
I did try making some window to configure it during key creation, but I rediscovered how much I suck at windows forms.
What I was thinking was to have 3 tabs: one with some very general options like you describe, one with all of the methods that requires scopes, and one with all of the scopes. Then when you check/uncheck something, it would affect the data shown in the other tabs, so that you can always go see what scopes will be requested, before actually opening the auth page. That will also help making sure that you expose the correct scopes on the developer site, if you're really strict about what you want to configure there.

**Comment by wbSD** [comment]
I pulled myself together and have something coming with the checkbox window. Might have time to finish it tomorrow.
I'll make sure to have the simple version and some reasonable defaults.

**Comment by wbSD** [comment]
Okay, there it is!
There's now a button below the "Log in with EVE online" button called "Configure scopes", where you can enable/disable all the scopes to your heart's desire.
image

There are 3 tabs:
1. Simple, with the options you listed. These can be easily modified through the corresponding CCPAPIMethodsEnum.
image

2. Advanced, showing all of the character and corporation methods that require ESI scopes
image

3. ESI Scopes, showing all of the scopes that are used by methods in the Advanced tab
image

All checkboxes on all 3 tabs live updates the checkboxes on the other tabs, and the ones on the first tab have ThreeState enabled, meaning they show as partially checked if not all the related methods/scopes are allowed.
image
I have to be honest that some of the code is run multiple times. For example when you check/un-check Core Features, it loops through the same checks 16 times (twice per method). The Simple tab checkboxes also cycle through all 3 states when you check them manually, so you might have to click one twice to disable it.
All in all I'm pretty satisfied with the result, though.
Let me know what you think.

**Comment by wbSD** [comment]
I noticed that applying the "Header" attribute to an API method made it appear not only in the throbber context menu, but also in the Updates page in settings. Many of the methods weren't used in a CharacterQueryMonitor though, so it didn't make sense to show them in either.
Instead I moved the header+description into the ESIMethod attribute, so that the text can still be shown when configuring the ESI key scopes.

**Comment by peterhaneve** [comment]
Yeah I noticed that too, anything with a header becomes configurable in the settings, and if it never gets its own query monitor (e.g. implants, attributes) users will probably wonder why the change does not affect anything.

**Comment by wbSD** [comment]
I haven't yet tested the Attribute/Implant import problem with the new 304 NotModified responses, but I am pretty sure it's going to cause some problems.

I will look into removing the attribute/implant query changes from this PR and then it can be revisited at a later time.
The main reason I made it back into a query monitor was so that it could use the HasAccess check of the CharacterQueryMonitor (which checks on the access mask of the ESI key).
With the way the queries are currently linked on master, combined with the AccessMask changes of this PR, if you don't have access to Implants then Attributes will never be queried, as that happens in the Implants callback.

**Comment by wbSD** [comment]
After testing it thoroughly, this PR just has the exact same side effect as described in the original comment.
What isn't working is 4.0.5 (and the current master branch).

If you change your implants (and both implants and attributes have been queried/cached already), the implant query monitor will at some point query again and import the new implants. After that it loads attributes, but since the cache timer is longer on those, it doesn't get the new data.
Now it's stuck and will never get the new attributes. Every time implants are queried, it gets 304 and doesn't invoke the onSuccess action, which means attributes aren't queried again.

**Comment by wbSD** [comment]
This is now compatible with PKCE authentication as well.

File: src/EVEMon.Common/Enumerations/CCPAPI/ESIAPICorporationMethods.cs [hunk 1]
```diff
@@ -100,66 +112,79 @@ public enum ESIAPICorporationMethods : ulong
...
         /// Corporation member information.
         /// </summary>
+        [ESIMethod("/v1/corporations/{0:D}/membertracking/", Scope = "esi-corporations.track_members.v1")]
         CorporationMemberTracking = 1 << 15,
-        
+
         /// <summary>
         /// List of all outposts controlled by the corporation.
         /// </summary>
+        [ESIMethod("/v1/corporations/{0:D}/outposts/", Scope = "esi-corporations.read_outposts.v1")]
```
Comment (marfenij) [inline src/EVEMon.Common/Enumerations/CCPAPI/ESIAPICorporationMethods.cs]: there is no outpost API
https://esi.evetech.net/ui/?datasource=tranquility&version=latest#/
and now that scope `esi-corporations.read_outposts.v1` not supported

Comment (wbSD) [reply src/EVEMon.Common/Enumerations/CCPAPI/ESIAPICorporationMethods.cs]: On one of my other pull requests https://github.com/peterhaneve/evemon/pull/88 @peterhaneve said that a rewrite of the ESI frontend wasn't in the cards for now, so I am not going to spend time fixing this up until then.

**Pull request closed without merging**